### ISSUE TITLE
Autovacuum not reclaiming disk space

### ISSUE THREAD

**Issue by zregvart** [syndesisio/syndesis#3922]
## This is a...

<pre><code>
[ ] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Documentation issue or request
</code></pre>

## Description

We had a Syndesis instance (1.4.10, 76bc7c6f389f65ba0518bdd05823d247f4870a90) running for ~3 days (2 days 17 hours) with 6 integrations generating 730K messages.

The `syndesis-db` (1GiB) volume was on the verge of running out of disk space (12MiB left). And `VACUUM FULL jsondb` could not be run:

```
ERROR:  could not extend file "base/16385/16845": No space left on device
HINT:  Check free disk space.
STATEMENT:  VACUUM FULL ANALYSE jsondb;
```

Fetching activities resulted in 504 (Gateway timeout).

With ~30K rows in `jsondb` table the overwhelming majority of disk space was taken up with deleted rows. That were not vacuumed by autovacuum.

The only way to get the stream of activities was to copy the table content to a new table, drop the `jsondb` table, rename the new table to `jsondb` and recreate indexes.

i.e.
```
> SELECT * INTO jsondb2 FROM jsondb;
> DROP TABLE jsondb;
> ALTER TABLE jsondb2 RENAME TO jsondb;
> CREATE UNIQUE INDEX jsondb_pkey ON jsondb USING btree (path);
> CREATE INDEX jsondb_idx ON jsondb USING btree (idx, value) WHERE (idx IS NOT NULL)
```

I think we should try to tune the `autovacuum` so that the disk space is reclaimed (autovacuum will never perform `VACUUM FULL`) or bring back the DB maintenance job to perform `VACUUM FULL`.

The level of severity depends on how we envision the users using Syndesis, if this level of activity is acceptable (looks to me like it should be), then this feels like `p0`.

**Comment by zregvart** [comment]
From the logs on Saturday, logs don't have timestamps, but they are rotated so this is the last entry on Saturday (was running from Friday to Monday):

```
LOG:  automatic vacuum of table "syndesis.public.jsondb": index scans: 0
	pages: 0 removed, 72236 remain, 1996 skipped due to pins
	tuples: 0 removed, 855896 remain, 797779 are dead but not yet removable
	buffer usage: 89805 hits, 70138 misses, 359 dirtied
	avg read rate: 4.438 MB/s, avg write rate: 0.023 MB/s
	system usage: CPU 0.55s/0.25u sec elapsed 123.46 sec
```

Key is `797779 are dead but not yet removable` which leaves 780K tuples unclaimed in the data files, and `elapsed 123.46 sec` which is a really long time for the vacuum to perform.

I've changed the `autovacuum_vacuum_cost_limit` from `1000` to `2000` and it seems good thus far, here it is running for about 10 hours:

```
LOG:  automatic vacuum of table "syndesis.public.jsondb": index scans: 1
        pages: 0 removed, 1859 remain, 0 skipped due to pins
        tuples: 2915 removed, 22780 remain, 0 are dead but not yet removable
        buffer usage: 2611 hits, 431 misses, 6 dirtied
        avg read rate: 81.327 MB/s, avg write rate: 1.132 MB/s
        system usage: CPU 0.00s/0.01u sec elapsed 0.04 sec
```

Lets see after a few days.

**Comment by zregvart** [comment]
After ~14h of running with the changes in #3938 we are still facing the same issue:

```
$ echo "select idx_scan, idx_tup_fetch, n_tup_ins, n_tup_upd, n_tup_del, n_tup_hot_upd, n_live_tup, n_dead_tup, n_mod_since_analyze, last_autovacuum, last_autoanalyze, vacuum_count, autovacuum_count from pg_stat_all_tables where relname = 'jsondb'" | psql -U syndesis
 idx_scan | idx_tup_fetch | n_tup_ins | n_tup_upd | n_tup_del | n_tup_hot_upd | n_live_tup | n_dead_tup | n_mod_since_analyze |        last_autovacuum        |       last_autoanalyze        | vacuum_count | autovacuum_count 
----------+---------------+-----------+-----------+-----------+---------------+------------+------------+---------------------+-------------------------------+-------------------------------+--------------+------------------
   457627 |     181753252 |    249920 |    200498 |    222484 |         60121 |      24823 |     371436 |                6257 | 2018-10-25 11:18:50.228729+00 | 2018-10-25 10:53:19.475484+00 |            0 |             3091
```

That's `371436` tuples that are not reclaimed.

Disk space is dwindling:
```
$ df -kh /var/lib/pgsql/data/
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme3n1    976M  291M  618M  33% /var/lib/pgsql/data
```

autovacuum is running:
```
2018-10-25 11:21:52 UTC   00000 5bd1a6f7.125eb LOG:  automatic vacuum of table "syndesis.public.jsondb": index scans: 0
        pages: 0 removed, 19582 remain, 3 skipped due to pins
        tuples: 0 removed, 396275 remain, 371435 are dead but not yet removable
        buffer usage: 27801 hits, 17735 misses, 51 dirtied
        avg read rate: 1.567 MB/s, avg write rate: 0.005 MB/s
        system usage: CPU 0.08s/0.07u sec elapsed 88.39 sec
```

It is unable to remove the tuples as there are transactions that hold locks over those rows:

```
$ echo "SELECT query, state,locktype,mode FROM pg_locks JOIN pg_stat_activity USING (pid) WHERE relation::regclass = '"'"'jsondb'"'"'::regclass AND granted IS TRUE AND backend_xmin IS NOT NULL;" | psql -d syndesis
                                                         query                                                         |        state        | locktype |           mode           
-----------------------------------------------------------------------------------------------------------------------+---------------------+----------+--------------------------
 INSERT into jsondb (path, value, ovalue) values ($1, $2, $3) ON CONFLICT (path) DO UPDATE SET value = $4, ovalue = $5 | active              | relation | RowExclusiveLock
 INSERT into jsondb (path, value, ovalue) values ($1, $2, $3) ON CONFLICT (path) DO UPDATE SET value = $4, ovalue = $5 | active              | relation | AccessShareLock
 select path,value,ovalue from jsondb where path LIKE $1 order by path ASC                                             | active              | relation | AccessShareLock
 select path,value,ovalue from jsondb where path LIKE $1 order by path ASC                                             | idle in transaction | relation | AccessShareLock
 autovacuum: VACUUM public.jsondb                                                                                      | active              | relation | ShareUpdateExclusiveLock
(5 rows)
```

On the server there are two threads accessing the database:
```
"Logs Controller Scheduler [running]: pollPods" #30 prio=5 os_prio=0 tid=0x00007ff6e9a585e0 nid=0xb6 runnable [0x00007ff6ec048000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:140)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:109)
	at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:67)
	at org.postgresql.core.PGStream.receiveChar(PGStream.java:280)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1916)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:288)
	- locked <0x00000000e7f93180> (a org.postgresql.core.v3.QueryExecutorImpl)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:430)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:356)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:168)
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:157)
	at sun.reflect.GeneratedMethodAccessor114.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tomcat.jdbc.pool.StatementFacade$StatementProxy.invoke(StatementFacade.java:114)
	at com.sun.proxy.$Proxy158.execute(Unknown Source)
	at org.skife.jdbi.v2.SQLStatement.internalExecute(SQLStatement.java:1327)
	at org.skife.jdbi.v2.Query.iterator(Query.java:240)
	at io.syndesis.server.jsondb.impl.SqlJsonDB.getAsStreamingOutput(SqlJsonDB.java:269)
	at io.syndesis.server.jsondb.JsonDB.getAsStream(JsonDB.java:160)
	at io.syndesis.server.jsondb.JsonDB.getAsByteArray(JsonDB.java:130)
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController.dbGet(ActivityTrackingController.java:330)
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController.dbGet(ActivityTrackingController.java:326)
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController.pollPods(ActivityTrackingController.java:273)
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController$$Lambda$90/1605950465.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"Logs Controller [running]: processEventQueue" #29 prio=5 os_prio=0 tid=0x00007ff6e9ef1930 nid=0xb5 runnable [0x00007ff6ec081000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:140)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:109)
	at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:67)
	at org.postgresql.core.PGStream.receiveChar(PGStream.java:280)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1916)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:460)
	- locked <0x00000000e7f88f40> (a org.postgresql.core.v3.QueryExecutorImpl)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:793)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1659)
	at sun.reflect.GeneratedMethodAccessor170.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tomcat.jdbc.pool.StatementFacade$StatementProxy.invoke(StatementFacade.java:114)
	at com.sun.proxy.$Proxy158.executeBatch(Unknown Source)
	at org.skife.jdbi.v2.PreparedBatch.internalBatchExecute(PreparedBatch.java:196)
	at org.skife.jdbi.v2.PreparedBatch.execute(PreparedBatch.java:100)
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController.lambda$writeBatch$2(ActivityTrackingController.java:201)
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController$$Lambda$414/308657533.inTransaction(Unknown Source)
	at org.skife.jdbi.v2.tweak.transactions.LocalTransactionHandler.inTransaction(LocalTransactionHandler.java:183)
	at org.skife.jdbi.v2.BasicHandle.inTransaction(BasicHandle.java:338)
	at org.skife.jdbi.v2.DBI$5.withHandle(DBI.java:350)
	at org.skife.jdbi.v2.DBI.withHandle(DBI.java:298)
	at org.skife.jdbi.v2.DBI.inTransaction(DBI.java:346)
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController.writeBatch(ActivityTrackingController.java:168)
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController.processEventQueue(ActivityTrackingController.java:374)
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController$$Lambda$88/568239889.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

I don't see a way for the autovacuum to work in such case the only solution I can think of is to run `VACUUM FULL` periodically and while doing so block access to the table.

**Comment by zregvart** [comment]
I just ran `VACUUM FULL` and that didn't help, the amount of dead tuples is still ~380K.

**Comment by zregvart** [comment]
I'm noticing that there `idle in transaction` connections:

```
1001120+   1966  0.0  0.1 171480 26852 ?        Ss   Oct24   0:07 postgres: syndesis syndesis 10.1.0.1(44962) idle in transaction
1001120+   1967  2.0  0.4 171612 74720 ?        Ss   Oct24  18:55 postgres: syndesis syndesis 10.1.0.1(44964) idle in transaction
```

It could be that we're missing a `COMMIT` or are not releasing connections to the pool.

A restart of the server pod could point to that, if it solves the issue.

**Comment by zregvart** [comment]
Yeah I don't see any threads using the database still the two `idle in transaction` connection remains. I'm going to restart the server pod and see if the autovacuum clears up the dead tuples.

**Comment by zregvart** [comment]
Immediately after the restart autovacuum removed the dead tuples:

```
2018-10-25 11:39:53 UTC   00000 5bd1ab80.13f69 LOG:  automatic vacuum of table "syndesis.public.jsondb": index scans: 1
        pages: 0 removed, 21034 remain, 0 skipped due to pins
        tuples: 379680 removed, 24769 remain, 0 are dead but not yet removable
        buffer usage: 64363 hits, 35737 misses, 42444 dirtied
        avg read rate: 33.605 MB/s, avg write rate: 39.911 MB/s
        system usage: CPU 0.30s/0.18u sec elapsed 8.30 sec
```

**Comment by chirino** [comment]
Any thoughts on which threads are using those connections?

**Comment by zregvart** [comment]
What I find puzzling is if we have a connection leak, we would see it exhausting the connection pool, we don't see that after the server pod restart. I think the autovacuum was prevented by `ROW EXCLUSIVE` lock taken by `INSERT into jsondb (path, value, ovalue) values ($1, $2, $3) ON CONFLICT (path) DO UPDATE SET value = $4, ovalue = $5`, that would mean that that `INSERT` took a really long time, not sure what are the circumstances of that. Perhaps the issue is that we have two `INSERT` statements run in parallel, hard to tell.

**Comment by zregvart** [comment]
How about setting query timeout via `Statement::setQueryTimeout`[1]

[1] https://docs.oracle.com/javase/8/docs/api/java/sql/Statement.html#setQueryTimeout-int-

**Comment by zregvart** [comment]
I would summarise this as:

An `ROW EXCLUSIVE` lock on the `jsondb` table prevented autovacuum from running which resulted in high number of dead tuples (371436), the statement causing the lock was an `INSERT` statement issued by the activity logging logic. If left over time this would have filled up the available disk space and be detrimental to database performance.

The issue was resolved by deleting the `syndesis-server` pod, thereby restarting the Java process trying to perform that `INSERT` which allowed autovacuum to remove the dead tuples.

In the 7 days since this has not reoccurred and I've found no way of reproducing the issue.

I propose that we close this with the resolution of _won't fix_ and work towards alternate activity logging storage solution.

**Comment by KurtStam** [comment]
+1 to won't fix. But let's open a new issue that references this one that talks about alternative logging using maybe elastic. We've been abusing the DB long enough for this.

Would a quick intermediate step be to use a different table for logging (not the jsondb) and explicitly set no locking on the insert.
